orderbook/src/database.rs

```diff
@@ -4,13 +4,46 @@ mod orders;
 mod trades;
 
 use anyhow::Result;
+use chrono::{DateTime, Utc};
+use contracts::gpv2_settlement::Event as ContractEvent;
+use ethcontract::Event as EthContractEvent;
+use futures::{stream::BoxStream, StreamExt};
+use model::{
+    order::{Order, OrderUid},
+    trade::Trade as ModelTrade,
+};
+use shared::event_handling::EventIndex;
 use sqlx::{Executor, PgPool, Row};
 use std::collections::HashMap;
 
 pub use events::*;
 pub use orders::OrderFilter;
 pub use trades::TradeFilter;
 
+#[async_trait::async_trait]
+pub trait Database: Send + Sync {
+    async fn clear(&self) -> Result<()>;
+    async fn count_rows_in_tables(&self) -> Result<HashMap<&'static str, i64>>;
+
+    async fn block_number_of_most_recent_event(&self) -> Result<u64>;
+    async fn insert_events(&self, events: Vec<(EventIndex, Event)>) -> Result<()>;
+    async fn replace_events(
+        &self,
+        delete_from_block_number: u64,
+        events: Vec<(EventIndex, Event)>,
+    ) -> Result<()>;
+    fn contract_to_db_events(
+        &self,
+        contract_events: Vec<EthContractEvent<ContractEvent>>,
+    ) -> Result<Vec<(EventIndex, Event)>>;
+
+    async fn insert_order(&self, order: &Order) -> Result<(), InsertionError>;
+    async fn cancel_order(&self, order_uid: &OrderUid, now: DateTime<Utc>) -> Result<()>;
+    fn orders<'a>(&'a self, filter: &'a OrderFilter) -> BoxStream<'a, Result<Order>>;
+
+    fn trades<'a>(&'a self, filter: &'a TradeFilter) -> BoxStream<'a, Result<ModelTrade>>;
```

Comment:
If these particular order and trade related methods weren't directly on the database we could also implement this trait for `BalancerPoolStorage`.

Reply:
I am not completely sure what you mean. This trait is roughly the set of all operations we want do with the postgres db. So if we were going to store balancer related information in the db then we could add methods here or there could be a new trait for these balancer operations that is implemented by `Postgres` like our existing `MinFeeStoring`.

Reply:
All the trait methods you have declared on `Database` (except the ones I have highlighted) here are also defined for `BalancerPool` which is also a kind of in-memory "Database". Perhaps it would be worth stripping these less generic trait methods from the base Database trait and making an `OrderbookDatabase` which implements database as well as the orderbook specific methods. This way, we could also use the Database trait for Balancer `PoolStorage`. 

not necessary at all, just an idea.